### PR TITLE
Organic: make oscillator phase offsets per-note

### LINKED ISSUE #1499: Organic glitches
While trying some presets, I came to organic ones and I saw a little problem with all of them.

At the start and at the end of a note there is a glitch, a little click.
What's worse is that you can't avoid it because even if you play with the envelope you will still hear it when playing a note just after or during another one.

I exported some score with a 1osc preset, so without actually doing additive synthesis, let's say, and I opened it with Audacity.

Well, here is the problem
screenshot_9

The two channels have different phases, and in orther to achieve this feature, they start with random initial phase.

I'm not really into synthesis yet, but I do know that starting a sound with an initial phase other than 0 will lead to a click, or glitch.

So, do we need different phases for the two channels?
If yes, what about having it by letting one of the two channels start a bit after the first? Like this:

screenshot_10

Not sure if this is a bug, but it's not like I've heard this problem before now...

### PR DESCRIPTION
**Pull request by softrabbit**
Fixes some glitches mentioned in #1499 by creating an array on phase offsets for each note instead of the earlier array that was shared by the entire plugin instance and re-randomized on every note start.

### CODE REVIEW

**Comment by softrabbit** [comment]
I won't merge this on my own, as I have to admit my C++ isn't that strong. So, if anyone has objections or opinions on how to do this better, please chime in.

**Comment by tresf** [comment]
Only criticism i have is that `phaseOffsetLeft` and `phaseOffsetRight` are declared twice (once as member variables, once as local variables).  Doesn't this throw an used variable warning on compile or am I missing something?

**Comment by tresf** [comment]
Nevermind, that's the struct declaration I was looking at. :smile:

File: plugins/organic/organic.h [hunk 1]
```diff
@@ -153,6 +153,8 @@ public slots:
 		MM_OPERATORS
 		Oscillator * oscLeft;
 		Oscillator * oscRight;
+		float * phaseOffsetLeft;
+		float * phaseOffsetRight;
```
Comment (michaelgregorius) [inline plugins/organic/organic.h]: I'd propose to rather use an `std::vector<float>` here. So something like this:
```C++
std::vector<float> phaseOffsetLeft;
std::vector<float> phaseOffsetRight;
```
This saves some allocations on the heap and also the array deletes starting on line 333 in the implementation file.

**Comment by michaelgregorius** [comment]
I have added a line comment concerning the implementation. Other than that if this fixes the problem described in #1499 I'd propose to rebase, squash and merge into the master as #1499 is part of the 1.2 milestones.

**Comment by softrabbit** [comment]
@michaelgregorius, I only had a cursory read around the interwebs yet, but wouldn't a ```std::vector``` still allocate its storage on the heap, and ```std::array``` would be more appropriate?

**Comment by michaelgregorius** [comment]
@softrabbit You're right in that an `std::vector` might allocate the underlying storage on the heap. However, what really happens behind the scenes also depends on the implementation of the standard library. As you can see [here](http://en.cppreference.com/w/cpp/container/vector) the constructor of the `std::vector` takes a second argument which defaults to `std::allocator<T>`. The standard allocator might for example take the memory from an underlying pool so that there is not an allocation for each `vector` that needs memory. It's also possible to provide an own implementation of the allocator.

I had to read up on the `std::array` and it seems that it has the performance of raw C arrays with a certain amount of abstraction and I think it might also be a good fit for this case.

**Comment by zonkmachine** [comment]
> I won't merge this on my own, as I have to admit my C++ isn't that strong.

Same position here. I've tested this and vote merge.

**Comment by softrabbit** [comment]
I finally dug into the change discussed, and... 
* gcc 5.2 complained that ```std::array``` was experimental and needed a compiler option to work, so that might not be a good idea yet?
* I'm not sure of how to put ```std::vector``` to use in such a way that every note gets its own phase vectors, and every ```Oscillator``` created in ```OrganicInstrument::playNote()``` gets a pointer to a float in the vector, while keeping said pointer valid for the duration of the note, without going the new/delete route that's supposed to be eliminated by this juggling.

I'm tempted to merge and let someone else make it pretty. :confused:

**Comment by zonkmachine** [comment]
> I'm tempted to merge and let someone else make it pretty. :confused:

:+1:

**Comment by Umcaruje** [comment]
:+1:

**Pull request merged**